woocommerce/woocommerce-ios
# [Stats] Prevent crash when parsing a date as Daylight Saving Time begins

Closes: #3888


## Description


When parsing dates for `OrderStatsV4Interval` intervals, the parsing can fail when the date string matches the date and time that Daylight Saving Time (DST) begins for the provided time zone.

My understanding is this happens because the provided date/time doesn't exist (because of DST the time jumps forward to the next hour), and without a calendar, the formatter can't calculate the correct `Date`.

This is fixed by defining the calendar used to parse the date strings for `OrderStatsV4Interval`. 

## Testing instructions


You can observe the crash by removing the calendar from `OrderStatsV4Interval+Date.swift:25` and running the new test `OrderStatsV4Interval_DateTests.test_date_start_and_date_end_parsed_when_daylight_saving_time_begins()`.

Then, confirm using the calendar makes the test pass.

---
